### PR TITLE
Conversion to Gradle KTS

### PR DESCRIPTION
**Pull request by Doomsdayrs**
Convert Android Groovy Script to Kotlin Script.

As to possibly facilitate #33742 and or to simply match current Android Standards.

These are history destructive commits, as git will not recognize such heavy modifications as file moves.

All Logic is preserved per each commit sequentially. This means the first commit to the last preserves logic as to allow the first commit to be merged without the following ones, but not as to allow the following commits to be merged without the prior commit. 

---

- [x] `./mach build --debug --target x86_64-linux-android` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [ ] These changes fix #___ (GitHub issue number if applicable)

- [x] There are tests for these changes OR
- [ ] These changes do not require tests because ___

### CODE REVIEW

File: support/android/apk/servoview/build.gradle.kts [hunk 1]
```diff
@@ -0,0 +1,259 @@
...
+        }
+    }
+
+    var result = ""
+
+    ResourceGroovyMethods.eachFileRecurse(path, FileType.FILES, closureOf<File> {
+        if (this.name == filename) {
+            result = this.absolutePath
+        }
+    })
```
Comment (Doomsdayrs) [inline support/android/apk/servoview/build.gradle.kts:254]: This does not work, trying to find a replacement.

Comment (Doomsdayrs) [reply support/android/apk/servoview/build.gradle.kts:254]: Maybe it did work, the replacement is also failing to work though..

Comment (Doomsdayrs) [reply support/android/apk/servoview/build.gradle.kts]: Figured it out, `if (path.isNotBlank())` was missing.

**Comment by Doomsdayrs** [comment]
Alright, with that all the build code is now in Kotlin Script.

Next steps, Kotlinifying the various functions to no longer require Groovy Closures.

**Comment by Doomsdayrs** [comment]
I personally feel that I cannot improve this any further currently.

**Comment by Doomsdayrs** [comment]
Just found out that github butchered verifications when rebasing.

When I get home I'll repush to fix the signing.

**Comment by Doomsdayrs** [comment]
May someone enable the android CI for this PR?

**Comment by mrobinson** [comment]
```
Daemon will be stopped at the end of the build 
> Task :buildSrc:checkKotlinGradlePluginConfigurationErrors
> Task :buildSrc:pluginDescriptors
> Task :buildSrc:processResources NO-SOURCE
> Task :buildSrc:compileKotlin
> Task :buildSrc:compileJava NO-SOURCE
> Task :buildSrc:compileGroovy NO-SOURCE
> Task :buildSrc:classes UP-TO-DATE

> Task :buildSrc:jar
:jar: No valid plugin descriptors were found in META-INF/gradle-plugins

> Configure project :servoapp
w: file:///home/runner/work/servo/servo/support/android/apk/servoapp/build.gradle.kts:153:69: 'capitalize(): String' is deprecated. Use replaceFirstChar instead.

WARNING: API 'splits.density' is obsolete.
FAILURE: Build failed with an exception.

* Where:
Build file '/home/runner/work/servo/servo/support/android/apk/servoapp/build.gradle.kts' line: 55

* What went wrong:
SigningConfig with name 'release' not found.
```

**Comment by Doomsdayrs** [comment]
Ah, I know the solution for that.

(using "named" instead of "register" in build script)

**Comment by Doomsdayrs** [comment]
May someone try the android run again?

**Comment by Doomsdayrs** [comment]
Ah, Google play requires API 33 or higher. A tomorrow task, for now I sleep.

**Comment by Doomsdayrs** [comment]
Request for a test again.

**Comment by Doomsdayrs** [comment]
Touched the sun with that.

**Comment by Doomsdayrs** [comment]
Hopefully the final attempt.

**Comment by Doomsdayrs** [comment]
😮‍💨 Alright, so it wasn't because of API 34 being the target.

the solution is to correct the notification code, which I don't want to do due to maintaining scope of this PR, but seems I have to.

**Comment by Doomsdayrs** [comment]
I went out of scope and fixed lint issues as well. Hopefully now an android build will work! (it did on my PC before too, but :shrug: )

**Comment by Doomsdayrs** [comment]
PRAISE BE!

**Comment by mukilan** [comment]
Great work!

I was testing this change locally and it crashes on my device. It seems like the APK produced doesn't contain the native shared objects (libservoshell.so and libc++_shared.so). The [artifacts generated](https://github.com/servo/servo/actions/runs/11412144557) from the recent try run also confirm this - the APK is too small and doesn't have the 'lib' folder inside it.

I can investigate the issue later this week, but if you wanted to look into it, I suspect [this logic to copy the native objects](https://github.com/servo/servo/blob/4d11b2dc84d53b90ca5f2b729ffa1a5db2da45e9/support/android/apk/servoview/build.gradle#L144) is not being triggered as expected, perhaps due to a change in the name of the task we are looking to match using the regex.

**Comment by Doomsdayrs** [comment]
> Great work!
> 
> I was testing this change locally and it crashes on my device. It seems like the APK produced doesn't contain the native shared objects (libservoshell.so and libc++_shared.so). The [artifacts generated](https://github.com/servo/servo/actions/runs/11412144557) from the recent try run also confirm this - the APK is too small and doesn't have the 'lib' folder inside it.
> 
> I can investigate the issue later this week, but if you wanted to look into it, I suspect [this logic to copy the native objects](https://github.com/servo/servo/blob/4d11b2dc84d53b90ca5f2b729ffa1a5db2da45e9/support/android/apk/servoview/build.gradle#L144) is not being triggered as expected, perhaps due to a change in the name of the task we are looking to match using the regex.

Something something `tasks.forEach` is occurring before the merge task is completed. Something something now have to figure out where it goes. I am investigating this right now.

**Comment by Doomsdayrs** [comment]
Maestro, another build?

File: support/android/apk/buildSrc/src/main/kotlin/Interop.kt [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+import java.io.File
+import java.util.Locale
+import java.util.Properties
+
+fun Project.getTargetDir(debug: Boolean, arch: String): String {
+    val basePath = project.rootDir.parentFile.parentFile.parentFile.absolutePath
+    return basePath + "/target/android/" + getSubTargetDir(debug, arch)
+}
+
+fun Project.getNativeTargetDir(debug: Boolean, arch: String) {
```
Comment (Doomsdayrs) [inline support/android/apk/buildSrc/src/main/kotlin/Interop.kt:12]: This does not return a string, thus the path ends up with a `Kotlin.Unit`

**Comment by Doomsdayrs** [comment]
Perchance let the CI spin again?

**Comment by Doomsdayrs** [comment]
It succeeded and also had a chunky artifact size, praise!

**Comment by mukilan** [comment]
Thanks! I'll review this by the end of the week.

File: support/android/apk/buildSrc/src/main/kotlin/Interop.kt [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+fun Project.getNativeTargetDir(debug: Boolean, arch: String): String {
+    val basePath = project.rootDir.parentFile.parentFile.parentFile.absolutePath
+    return basePath + "/target/" + getSubTargetDir(debug, arch)
+}
+
+fun getSubTargetDir(debug: Boolean, arch: String): String {
+    return getRustTarget(arch) + "/" + if (debug) "debug" else "release"
+}
+
+fun Project.getJniLibsPath(debug: Boolean, arch: String): String =
```
Comment (mukilan) [inline support/android/apk/buildSrc/src/main/kotlin/Interop.kt:21]: Why are some functions defined on the `Project` but not others?

Comment (mukilan) [inline support/android/apk/buildSrc/src/main/kotlin/Interop.kt:21]: Why are some functions defined on the Project but not others?

File: support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt [hunk 4]
```diff
@@ -0,0 +1,20 @@
...
+// XX is currently hardcoded to 00, but can be used to distinguish multiple builds within
+// the same day.
+//
+// TODO: check if this interferes with caching of local builds and add option to use
+// a static version.
+val today = Date()
+
+val versionCodeString: String = SimpleDateFormat("yyyyMMdd00").format(today)
+
+val generatedVersionCode = versionCodeString.toInt()
```
Comment (mukilan) [inline support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt:20]: Does Gradle guarantee that the logic in `buildSrc` will be run only once for both servoapp and servoview. My concern is that `servoapp` and `servoview` artifacts produced by a single build might not have the same version if that is not garaunteed.

File: support/android/apk/servoapp/src/main/AndroidManifest.xml [hunk 5]
```diff
@@ -26,6 +27,7 @@
               <category android:name="android.intent.category.DEFAULT" />
               <category android:name="android.intent.category.BROWSABLE" />
               <data android:scheme="http" />
+              <data android:host="*"/>
```
Comment (mukilan) [inline support/android/apk/servoapp/src/main/AndroidManifest.xml:30]: What is the motivation for this change? Was it not working previously?

File: support/android/apk/buildSrc/src/main/kotlin/Interop.kt (hunk 3 above)
Comment (Doomsdayrs) [reply support/android/apk/buildSrc/src/main/kotlin/Interop.kt:21]: Unlike in groovy in which scope is implicit and not clear, Kotlin requires explicit scoping.

`Project.` means this function is an extension to `Project` classes, thus has access to public fields of the `Project` class, and can only be called from a project object or scope.

I shall add documentation to the top of the file in this regard.

File: support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt (hunk 4 above)
Comment (Doomsdayrs) [reply support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt:20]: From what I understand from experience and [documentation](https://docs.gradle.org/current/userguide/sharing_build_logic_between_subprojects.html), and [member](https://discuss.gradle.org/t/buildsrc-vs-build-logic/46708/2) of the gradle team, Yes it is only produced once.

File: support/android/apk/servoapp/src/main/AndroidManifest.xml (hunk 5 above)
Comment (Doomsdayrs) [reply support/android/apk/servoapp/src/main/AndroidManifest.xml:30]: It was a part of a series of multiple commits addressing android lint errors, which in turn caused the CI to fail.

You may experiment on a local machine by dropping the commits and running the android lint task from gradle.

File: support/android/apk/buildSrc/src/main/kotlin/Interop.kt (hunk 3 above)
Comment (Doomsdayrs) [reply support/android/apk/buildSrc/src/main/kotlin/Interop.kt:25]: [786edc0](https://github.com/servo/servo/pull/33772/commits/786edc07ad24676396fc98815c44e7dc70d8c879)

File: support/android/apk/servoapp/src/main/AndroidManifest.xml (hunk 5 above)
Comment (Doomsdayrs) [reply support/android/apk/servoapp/src/main/AndroidManifest.xml:30]: Checking in, Have you done the experiment yet?

File: support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt (hunk 4 above)
Comment (mukilan) [reply support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt:20]: Thanks for the links! It does look like Gradle guarantees that the `buildSrc` is **built once** but I am not sure if that also implies if the built code **executes** once, globally across all sub-projects. In any case, we should be safe for now as the version code only has a granularity of a day, but this might be something we need to keep in mind if we ever extend it to include minutes, seconds or some other transient state.

File: support/android/apk/servoapp/src/main/AndroidManifest.xml (hunk 5 above)
Comment (mukilan) [reply support/android/apk/servoapp/src/main/AndroidManifest.xml:30]: Sorry about the delay. I did try removing `host` locally and it didn't fail the build for me. Do you have links to the logs from CI where it failed?

File: support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt (hunk 4 above)
Comment (Doomsdayrs) [reply support/android/apk/buildSrc/src/main/kotlin/VersionCode.kt:20]: because the version code is assigned to a val, we can assume it compiles into a constant int and said class is just imported by other projects.

I'll resolve this for now.

File: support/android/apk/servoapp/src/main/AndroidManifest.xml (hunk 5 above)
Comment (mukilan) [reply support/android/apk/servoapp/src/main/AndroidManifest.xml]: I did a [test run on the CI ](https://github.com/mukilan/servo/actions/runs/11625426495) without the host directives and it succeeded.  I'll update this PR to remove those. We can always add it back if we find it is needed.

**Comment by mukilan** [comment]
Great work! Thank you for working through all the issue with the migration!

**Review by mukilan — APPROVED** [review]
